### PR TITLE
Implement side effects

### PR DESCRIPTION
**Pull request by me4502**
Attached to https://github.com/EngineHub/WorldEdit/pull/547

If we decide to go with this, I'll add it to all the adapters

### CODE REVIEW

File: spigot_v1_15_R2/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/Spigot_v1_15_R2.java [hunk 1]
```diff
@@ -414,8 +452,8 @@ public void notifyAndLightBlock(Location position, BlockState previousType) {
         IBlockData oldData = ((CraftBlockData) BukkitAdapter.adapt(previousType)).getState();
         IBlockData newData = craftWorld.getHandle().getType(blockPosition);
 
-        //        craftWorld.getHandle().r(blockPosition); // Re-light
-        craftWorld.getHandle().notifyAndUpdatePhysics(blockPosition, null, oldData, newData, newData, 1 | 2); // Update
+        craftWorld.getHandle().getChunkProvider().getLightEngine().a(blockPosition); // server should do lighting for us
+        updateNeighbours(craftWorld.getHandle(), blockPosition, null, oldData, newData, newData, 1 | 2); // Update
```
Comment (octylFractal) [inline spigot_v1_15_R2/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/Spigot_v1_15_R2.java]: Can we extract this + the `setBlock` code to a common method?

File: spigot_v1_15_R2/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/Spigot_v1_15_R2.java [hunk 2]
```diff
@@ -318,6 +320,42 @@ public BaseBlock getBlock(Location location) {
...
+                world.b(blockposition, oldBlock, actualBlock);
+            }
+
+            if ((i & 2) != 0 && (!world.isClientSide || (i & 4) == 0) && (world.isClientSide || chunk == null || chunk.getState() != null && chunk.getState().isAtLeast(
+                    PlayerChunk.State.TICKING))) {
+                ((ChunkProviderServer) world.getChunkProvider()).flagDirty(blockposition);
+            }
+
+            if (!world.isClientSide && (i & 1) != 0) {
+                world.getType(blockposition.west()).doPhysics(world, blockposition.west(), oldBlock.getBlock(), blockposition, false);
```
Comment (octylFractal) [inline spigot_v1_15_R2/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/Spigot_v1_15_R2.java]: I feel like we can do a loop here to reduce copy/paste:
```java
for (BlockPosition b : ImmutableList.of(bp.east(), bp.west(), ...) {
  world.getType(b).doPhysics(world, b, oldBlock.getBlock(), bp, false);
}
```

There might even already be something to iterate all directions?

File: spigot_v1_15_R2/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/Spigot_v1_15_R2.java [hunk 3]
```diff
@@ -318,6 +320,42 @@ public BaseBlock getBlock(Location location) {
...
+                world.getType(blockposition.south()).doPhysics(world, blockposition.south(), oldBlock.getBlock(), blockposition, false);
+
+                if (newBlock.isComplexRedstone()) {
+                    world.updateAdjacentComparators(blockposition, newBlock.getBlock());
+                }
+            }
+
+            if ((i & 16) == 0) {
+                int j = i & -2;
+                oldBlock.b(world, blockposition, j);
```
Comment (octylFractal) [inline spigot_v1_15_R2/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/Spigot_v1_15_R2.java]: A comment on what these three methods are doing would be nice, unless this is all copied from somewhere, in which case that place should be referenced.

**Review by octylFractal — CHANGES_REQUESTED** [review]

File: spigot_v1_15_R2/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/Spigot_v1_15_R2.java (hunk 2 above)
Comment (me4502) [reply spigot_v1_15_R2/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/Spigot_v1_15_R2.java]: Yeah, this was done as I copy/pasted it from MC code. I'll do a big cleanup when we decide what to keep etc

**Pull request merged**